backends/ent/store.go

```diff
@@ -42,436 +45,402 @@ func (backend *Backend) Store(doc *sbom.Document, opts *storage.StoreOptions) er
+			if nodeID, ok := backend.ctx.Value(nodeIDKey{}).(string); ok {
+				newPurpose.SetNodeID(nodeID)
+			}
 
-		if nodeID, ok := backend.ctx.Value(nodeIDKey{}).(string); ok {
-			newPurpose.SetNodeID(nodeID)
+			builders = append(builders, newPurpose)
 		}
 
-		builders = append(builders, newPurpose)
-	}
+		err := tx.Purpose.CreateBulk(builders...).
+			OnConflict().
+			Ignore().
+			Exec(backend.ctx)
+		if err != nil && !ent.IsConstraintError(err) {
+			return fmt.Errorf("ent.Tool: %w", err)
+		}
 
-	err := tx.Purpose.CreateBulk(builders...).
-		OnConflict().
-		Ignore().
-		Exec(backend.ctx)
-	if err != nil && !ent.IsConstraintError(err) {
-		return fmt.Errorf("ent.Tool: %w", err)
+		return nil
 	}
-
-	return nil
 }
 
-func (backend *Backend) saveTools(tools []*sbom.Tool) error {
-	if backend.client == nil {
-		return fmt.Errorf("%w", errUninitializedClient)
-	}
+func (backend *Backend) saveTools(tools []*sbom.Tool) txFunc {
+	return func(tx *ent.Tx) error {
+		builders := []*ent.ToolCreate{}
 
-	tx := ent.TxFromContext(backend.ctx)
-	builders := []*ent.ToolCreate{}
+		for _, t := range tools {
+			newTool := tx.Tool.Create().
+				SetName(t.Name).
+				SetVersion(t.Version).
+				SetVendor(t.Vendor)
 
-	for _, t := range tools {
-		newTool := tx.Tool.Create().
-			SetName(t.Name).
-			SetVersion(t.Version).
-			SetVendor(t.Vendor)
+			if metadataID, ok := backend.ctx.Value(metadataIDKey{}).(string); ok {
+				newTool.SetMetadataID(metadataID)
+			}
 
-		if metadataID, ok := backend.ctx.Value(metadataIDKey{}).(string); ok {
-			newTool.SetMetadataID(metadataID)
+			builders = append(builders, newTool)
 		}
 
-		builders = append(builders, newTool)
-	}
-
-	err := tx.Tool.CreateBulk(builders...).
-		OnConflict().
-		Ignore().
-		Exec(backend.ctx)
-	if err != nil && !ent.IsConstraintError(err) {
-		return fmt.Errorf("ent.Tool: %w", err)
-	}
-
-	return nil
-}
-
-func (backend *Backend) newNodeCreate(n *sbom.Node) *ent.NodeCreate {
-	tx := ent.TxFromContext(backend.ctx)
-
-	newNode := tx.Node.Create().
-		SetID(n.Id).
-		SetAttribution(n.Attribution).
-		SetBuildDate(n.BuildDate.AsTime()).
-		SetComment(n.Comment).
-		SetCopyright(n.Copyright).
-		SetDescription(n.Description).
-		SetFileName(n.FileName).
-		SetFileTypes(n.FileTypes).
-		SetLicenseComments(n.LicenseComments).
-		SetLicenseConcluded(n.LicenseConcluded).
-		SetLicenses(n.Licenses).
-		SetName(n.Name).
-		SetReleaseDate(n.ReleaseDate.AsTime()).
-		SetSourceInfo(n.SourceInfo).
-		SetSummary(n.Summary).
-		SetType(node.Type(n.Type.String())).
-		SetURLDownload(n.UrlDownload).
-		SetURLHome(n.UrlHome).
-		SetValidUntilDate(n.ValidUntilDate.AsTime()).
-		SetVersion(n.Version)
-
-	if nodeListID, ok := backend.ctx.Value(nodeListIDKey{}).(int); ok {
-		newNode.AddNodeListIDs(nodeListID)
-	}
-
-	return newNode
-}
+		err := tx.Tool.CreateBulk(builders...).
+			OnConflict().
+			Ignore().
+			Exec(backend.ctx)
+		if err != nil && !ent.IsConstraintError(err) {
```

Comment:
here we lose the error and return nil when `ent.IsConstraintError()` == true is this expected? (same on line 97)

Reply:
Yes, this is intentional; in most cases if we hit a unique constraint, we want to ignore it